### DISCUSSION TITLE
compiler for linux

### DISCUSSION THREAD

**Discussion by DeeDee22** [ooflet/Mi-Create#19] [Questions-EN] [answered]
Is there a chance for a compiler that works on Linux anytime soon? Or should I try to get a VM running? :-) (Or do you have a tip to make it work with wine?
Creating these watch faces really is the first time running Linux seems to be a problem :-D

**Comment by ooflet** [comment]
The compiler runs on wine perfectly fine so far, you need to run the compiler binary in src/compiler/compile.exe with wine and with parameters "-b {full path to .fprj} {output folder} {name of file.face} {id}". Id does not matter, you can set it to any number and the compiler simply ignores it.

The reason the compiler does not support Linux so far is because the maintainer is willing to compile for Linux. I'll try to get a Linux version of the compiler soon, expect before the final release of 1.0.

**Reply by DeeDee22** [reply]
oh, interesting. I tried to figure out the arguments from your code (not a python developer). I will try again. Do you happen to know what version of wine you are using? And did you have to install anything else?
I was wondering what the "0" as the last argument was :-)

**Reply by DeeDee22** [reply]
Btw: Thanks for making this project and answering my questions. I appreciate it.

**Reply by ooflet** [reply, marked as answer]
> oh, interesting. I tried to figure out the arguments from your code (not a python developer). I will try again. Do you happen to know what version of wine you are using? And did you have to install anything else? I was wondering what the "0" as the last argument was :-)

I use wine-9.13 from the Arch repository. The compiler uses the .NET framework and is written in C#, so you will need to install Wine Mono. Wine should automatically prompt you to install it when running the compiler.

I am also in the process of getting a Linux binary from the maintainer, so hang tight 😁

**Reply by DeeDee22** [reply]
I'll give it another try with a newer wine version, or otherwise I'll wait for the Linux binary :-)

**Comment by DeeDee22** [comment]
me again :-) the compiler seems to work with the new wine. at least it produces a .face file. Still trying to figure out how to get it onto my watch (I am using gadgetbridge). What I notice is that the .info file generated by the compiler will always say "Type="1" Size="1875887" DeviceType="1" Title="blabla" DeviceVersion="7.1.0" DeciceTypeName="Mi Band 7 Pro"" even when I chose a different device type when setting up the project. I checked the fprj file and the DeviceType there is correct (compared to the DeviceInfo.db file). Can this be a problem?

**Reply by ooflet** [reply]
It's always been an issue with the compiler, if you decompile the compiler (using something like dnspy) you can see that the .info file is a hardcoded string. It's there to prevent EasyFace from displaying a .info file not found. I'll probably make it delete the .info file after compiling.

I'm not sure if GadgetBridge is able to load up watch faces for RW3A because I have never used GadgetBridge. You can see the process that I use to upload watch faces at https://ooflet.github.io/docs/quickstart/testing

**Reply by DeeDee22** [reply]
ah, ok.
I can upload watch faces from here https://amazfitwatchfaces.com with gadgetbridge, so I was hoping it would work. :-)

**Reply by ooflet** [reply]
I just checked GadgetBridge and they appear to support watchface management for Xiaomi Protobuf watches https://gadgetbridge.org/basics/topics/xiaomi-protobuf/ (which include RW3A). It's way less cumbersome than getting access to Mi Fitness mod servers, so you might want to give it a shot!

EDIT: There are instructions on how to install watchfaces here: https://gadgetbridge.org/internals/features/installer/

**Reply by DeeDee22** [reply]
> I just checked GadgetBridge and they appear to support watchface management for Xiaomi Protobuf watches https://gadgetbridge.org/basics/topics/xiaomi-protobuf/ (which include RW3A). It's way less cumbersome than getting access to Mi Fitness mod servers, so you might want to give it a shot!
> 
> EDIT: There are instructions on how to install watchfaces here: https://gadgetbridge.org/internals/features/installer/

yeah, I tried that, it just kind of doesn´t do anything? So I suppose something might be wrong with my bin file. I'll try to mayb find something in a log file.

**Comment by DeeDee22** [comment]
ok, update: I used the unpack.exe to maybe find a problem with my .face file and it did indeed have some problem with some of the images on there (image processing err: ImageMagick.MagickCorruptImageErrorException: insufficient image data in file `Z:\home\geraldine\geeband\images\img_0001_2_0.bmp' @ error/bmp.c/ReadBMPImage/963)

When I removed the offending images I could successfully upload the watch face to my watch!!
I still need to figure out what is different about those images (couldn't find anyting in the xml file at first glance but this is progress. Love it! Thank you! :-)

Edit: The offending file is a .png file, just like the one that works. I have no idea why unpack.exe thinks it should be a .bmp. File. It acutally does unpack it into a bmp file containing a weirdly distorted version of the original.
